src/banman.cpp

```diff
@@ -39,9 +39,11 @@ BanMan::~BanMan()
 
 void BanMan::DumpBanlist()
 {
-    SweepBanned(); // clean unused entries (if bantime has expired)
-
-    if (!BannedSetIsDirty()) return;
+    {
+        LOCK(m_cs_banned);
```

Comment:
first commit: if there are any races possible, shouldn't this lock be held for the whole duration of the function?

Reply:
Yes. But another dedicated mutex should be used for that. No need to hold the `m_banned_mutex` during disk operations: `m_ban_db.Write(banmap)` and `LogPrint(BCLog::NET, ...)`.

Do you prefer to add it into this PR?

Reply:
Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

Reply:
> Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

I'm not sure if I understand your Q correctly. But if you mean that `m_ban_db.Write(banmap)` modifies `m_banned`, it is not the case because a local copy `banmap` is used.

And isn't it a suboptimal design when callers of `BanMan` public functions are waiting for a relatively and unpredictably long I/O operation?

Reply:
Setting the dirty flag to false can only be done after the map has been written, so I don't see how a second mutex can solve this

Reply:
#### Case A -- current implementation (both master and this PR)

Nothing prevents from `m_ban_db.Write(banmap)` concurrently from different threads.

#### Case B -- `BanMan::DumpBanlist()` acquires `m_cs_banned`/`m_banned_mutex` for the whole its duration

Performance degradation is possible. E.g., caller of `BanMan::IsDiscouraged()` must wait for `BanMan::DumpBanlist()` finished in another thread.

#### Case C -- `BanMan::DumpBanlist()` acquires a dedicated `m_dump_mutex` for the whole its duration

No concurrent `m_ban_db.Write(banmap)` calls. No performance degradation.

---
> Setting the dirty flag to false can only be done after the map has been written

How about the following patch:
```diff
--- a/src/banman.cpp
+++ b/src/banman.cpp
@@ -45,11 +45,12 @@ void BanMan::DumpBanlist()
         SweepBanned();
         if (!m_is_dirty) return;
         banmap = m_banned;
+        m_is_dirty = false;
     }
 
     int64_t n_start = GetTimeMillis();
-    if (m_ban_db.Write(banmap)) {
-        WITH_LOCK(m_banned_mutex, m_is_dirty = false);
+    if (!m_ban_db.Write(banmap)) {
+        WITH_LOCK(m_banned_mutex, m_is_dirty = true);
     }
 
     LogPrint(BCLog::NET, "Flushed %d banned node addresses/subnets to disk  %dms\n", banmap.size(),
```
?

Reply:
> How about the following patch:

Can you explain the patch?

Reply:
Currently, when thread A performs `m_ban_db.Write(banmap)` thread B can set `m_is_dirty = true` which will be overwritten by thread A just after successful finish of `m_ban_db.Write(banmap)` call. This creates inconsistency between the actual `m_banned` state and the `m_is_dirty` flag.

With the suggested patch it is guaranteed that every time we access `m_is_dirty` its value is consistent with the actual `m_banned` state.

Reply:
But what is the point of the dirty flag when it will effectively be always true?

Reply:
It's being set to `false` in `BanMan::DumpBanlist()`, it stays `false` when `m_ban_db.Write(banmap))` is successful. And it stays `false` if other member functions do no change `m_banned`.

It allow to skip `m_ban_db.Write(banmap))` for unchanged `m_banned`.

Reply:
So you are suggesting this patch combined with a dedicated dump mutex?

Reply:
Yes, but in another PR to keep this one focused on the transition form `RecursiveMutex` to `Mutex`.

Reply:
Strictly speaking, fixing the logical race is a bugfix, so it might be better to do it first to allow easier cherry-picking to older commits. Would you mind going with the fix first?

Reply:
Sure!

Reply:
> Is it that common to modify (write) the banmap to disk that this requires a separate mutex?

Please check out  `DumpMempool` implementation.

Reply:
> Would you mind going with the fix first?

Done in #24168.